### ISSUE TITLE
Problem with non-global variables within ?{} code in regexps

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8349]
Migrated from [rt.perl.org#38639](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=38639) (status was 'resolved')

Searchable as RT38639$

**Comment by p5pRT** [comment]
### From richard@highwayman.com

#### Created by richard@highwayman.com




The problem is in setting variable values in ?\{\} sections of a regexp

I am doing this within a foreach loop and the value is set the first time
around the loop\, but the second time the value is not set\.

Further experimentation has shown this works OK if the variable is global
\(declared at the outermost level\) but not if the variable is declared
within the foreach loop itself\.

Possibly it's not meant to work for non\-global variables\.\.\. in which case
the bug is that it works once :\) and doesn't fail every time around\.

Fairly simple demo program below&#8203;: the output I get is&#8203;:

C&#8203;:\\work\\scan\\logs>perl \-w test\.pl
Test 1
T = 'abc'
\-\-\-\-
T = 'abc'
\-\-\-\-
Test 2
S = 'abc'
\-\-\-\-
Use of uninitialized value in concatenation \(\.\) or string at test\.pl line 52\.
S = ''
\-\-\-\-

C&#8203;:\\work\\scan\\logs>


and here's the program that gives the effect\.  I've tried it on several
different versions of Perl\, both on Windows and Linux and get the same
effect on all of them :\(

Hope that's all clear \-\- if not then please just ask\.

Thanks

\-\-\-example program\-\-\-
\#
\# seems to be a Perl bug connected with the use of non\-global variables
\# in ?\{\} constructs within regexps
\#
\# Richard Clayton \<richard@&#8203;highwayman\.com>
\# 26 Feb 2006
\#
\# from the documentation I think the problem is the second time around
\# when the local variable isn't set \.\.\. but it may be that it's not
\# supposed to work\, in which case the first time around is the problem :\)
\#

use strict;

print "Test 1\\n";

\# first test \.\.\. the variable in the ?\{\} within the regexp is global

my @&#8203;a;

$a\[0\] = "foobar";
$a\[1\] = "foobar";

my $T;

foreach my $a \(@&#8203;a\)
\{
&nbsp;   if \($a =~ /^foobar\(?\{$T = "abc"\}\)$/\)
&nbsp;   \{
&nbsp;      print "T = '$T'\\n";
&nbsp;   \}

&nbsp;   print "\-\-\-\-\\n";
\}

print "Test 2\\n";

\# second test \.\.\. the variable in the ?\{\} within the regexp is local
\# and on the SECOND time around the loop it isn't set

my @&#8203;b;

$b\[0\] = "foobar";
$b\[1\] = "foobar";

foreach my $b \(@&#8203;b\)
\{
&nbsp;   my $S;

&nbsp;   if \($b =~ /^foobar\(?\{$S = "abc"\}\)$/\)
&nbsp;   \{
&nbsp;      print "S = '$S'\\n";
&nbsp;   \}

&nbsp;   print "\-\-\-\-\\n";
\}

\# end of example
\-\-\-end of example program\-\-

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.7:

Configured by builder at Wed Nov  2 08:44:18 2005.

Summary of my perl5 (revision 5 version 8 subversion 7) configuration:
  Platform:
    osname=MSWin32, osvers=5.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -Zi -DNDEBUG -O1 -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT -DNO_HASH_SEED -DUSE_SITECUSTOMIZE -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='12.00.8804', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"C:\apps\perl584\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"C:\apps\perl584\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    Iin_load_module moved for compatibility with build 806
    Avoid signal flag SA_RESTART for older versions of HP-UX
    PerlEx support in CGI::Carp
    Less verbose ExtUtils::Install and Pod::Find
    instmodsh upgraded from ExtUtils-MakeMaker-6.25
    Patch for CAN-2005-0448 from Debian with modifications
    Upgrade to Time-HiRes-1.76
    25774 Keys of %INC always use forward slashes
    25747 Accidental interpolation of $@ in Pod::Html
    25362 File::Path::mkpath resets errno
    25181 Incorrect (X)HTML generated by Pod::Html
    24999 Avoid redefinition warning for MinGW
    24699 ICMP_UNREACHABLE handling in Net::Ping
    21540 Fix backward-compatibility issues in if.pm


@INC for perl v5.8.7:
    c:/apps/Perl584/lib
    c:/apps/Perl584/site/lib
    .


Environment for perl v5.8.7:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=c:\apps\Perl584\bin\;c:\apps\batch;c:\apps\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\Program Files\ATI Technologies\ATI Control Panel;C:\PROGRA~1\COMMON~1\SONICS~1\;C:\Program Files\Microsoft Office\OFFICE11\Business Contact Manager\IM;C:\Program Files\Microsoft SQL Server\80\Tools\Binn\;C:\Program Files\Microsoft Office\OFFICE11\Business Contact Manager\;c:\cygwin\bin\;C:\altera\quartus41\bin;C:\altera\quartus41\eda\sim_lib\excalibur\swift\lib\pcnt.lib;C:\altera\quartus41\bin;c:\wince420\platform\EPXA1withDisplay\bin;C:\Program Files\Microsoft SDK\Bin\.;C:\Program Files\Microsoft SDK\Bin\WinNT\.;C:\Program Files\Microsoft Platform SDK for Windows XP SP2\Bin\.;C:\Program Files\Microsoft Platform SDK for Windows XP SP2\Bin\WinNT\.;C:\Program Files\MySQL\MySQL Server 4.1\bin;C:\Program Files\Network Associates\PGPNT;C:\Program Files\Microsoft Visual Studio\Common\Tools\WinNT;C:\Program Files\Microsoft Visual Studio\Common\MSDev98\Bin;C:\Program Files\Microsoft Visual Studio\Common\
    PERL_BADLANG (unset)
    SHELL (unset)


```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

> \[richard@&#8203;highwayman\.com \- Sun Feb 26 08&#8203;:06&#8203;:10 2006\]&#8203;:
> 
> 
> This is a bug report for perl from richard@&#8203;highwayman\.com\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.7\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> 
> The problem is in setting variable values in ?\{\} sections of a regexp
> 
> I am doing this within a foreach loop and the value is set the first
>    time
> around the loop\, but the second time the value is not set\.
> 
> Further experimentation has shown this works OK if the variable is
>    global
> \(declared at the outermost level\) but not if the variable is declared
> within the foreach loop itself\.
> 
> Possibly it's not meant to work for non\-global variables\.\.\. in which
>    case
> the bug is that it works once :\) and doesn't fail every time around\.
> 
> Fairly simple demo program below&#8203;: the output I get is&#8203;:
> 
> C&#8203;:\\work\\scan\\logs>perl \-w test\.pl
> Test 1
> T = 'abc'
> \-\-\-\-
> T = 'abc'
> \-\-\-\-
> Test 2
> S = 'abc'
> \-\-\-\-
> Use of uninitialized value in concatenation \(\.\) or string at test\.pl
>    line 52\.
> S = ''
> \-\-\-\-
> 
> C&#8203;:\\work\\scan\\logs>
> 
> 
> and here's the program that gives the effect\.  I've tried it on
>    several
> different versions of Perl\, both on Windows and Linux and get the same
> effect on all of them :\(
> 
> Hope that's all clear \-\- if not then please just ask\.
> 
> Thanks
> 
> \-\-\-example program\-\-\-
> \#
> \# seems to be a Perl bug connected with the use of non\-global
>    variables
> \# in ?\{\} constructs within regexps
> \#
> \# Richard Clayton \<richard@&#8203;highwayman\.com>
> \# 26 Feb 2006
> \#
> \# from the documentation I think the problem is the second time around
> \# when the local variable isn't set \.\.\. but it may be that it's not
> \# supposed to work\, in which case the first time around is the problem
>    :\)
> \#
> 
> use strict;
> 
> print "Test 1\\n";
> 
> \# first test \.\.\. the variable in the ?\{\} within the regexp is global
> 
> my @&#8203;a;
> 
> $a\[0\] = "foobar";
> $a\[1\] = "foobar";
> 
> my $T;
> 
> foreach my $a \(@&#8203;a\)
> \{
>    if \($a =~ /^foobar\(?\{$T = "abc"\}\)$/\)
>    \{
>       print "T = '$T'\\n";
>    \}
> 
>    print "\-\-\-\-\\n";
> \}
> 
> print "Test 2\\n";
> 
> \# second test \.\.\. the variable in the ?\{\} within the regexp is local
> \# and on the SECOND time around the loop it isn't set
> 
> my @&#8203;b;
> 
> $b\[0\] = "foobar";
> $b\[1\] = "foobar";
> 
> foreach my $b \(@&#8203;b\)
> \{
>    my $S;
> 
>    if \($b =~ /^foobar\(?\{$S = "abc"\}\)$/\)
>    \{
>       print "S = '$S'\\n";
>    \}
> 
>    print "\-\-\-\-\\n";
> \}
> 
> \# end of example
> \-\-\-end of example program\-\-
> 

This problem also occurs in bleadperl@&#8203;27717\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rspier

manually unmerged from 8027

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
